Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+struct HTTPRequestStateMachine {
+    fileprivate enum State {
+        case initialized
+        case running(RequestState, ResponseState)
+        case finished
+
+        case failed(Error)
+    }
+
+    fileprivate enum RequestState {
+        enum ExpectedBody {
+            case length(Int)
+            case stream
+        }
+
+        enum ProducerControlState: Equatable {
+            case producing
+            case paused
+        }
+
+        case verifyRequest
+        case streaming(expectedBodyLength: Int?, sentBodyBytes: Int, producer: ProducerControlState)
+        case endSent
+    }
+
+    fileprivate enum ResponseState {
+        enum StreamControlState {
+            case downstreamHasDemand
+            case readEventPending
+            case waiting
+        }
+
+        case initialized
+        case receivingBody(StreamControlState)
+        case endReceived
+    }
+
+    enum Action {
+        enum AfterHeadContinueWith {
```

Comment:
This needs a new name. Maybe:

```suggestion
        enum NextMessageToSend {
```